### ISSUE TITLE
404 renders __NEXT_DATA__ as string to DOM when using RSC after deploy, locally no response

### ISSUE THREAD

**Issue by KlotzJesse** [vercel/next.js#32515]
### What version of Next.js are you using?

12.0.8-canary.5 up to 12.0.8-canary.7

### What version of Node.js are you using?

local: 14.8.1 // vercel: 14

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

Vercel

### Describe the Bug

When updating to newest canaries, starting by 12.0.8-canary.5, using `serverComponents: true` and `concurrentFeatures: true` the 404 page (default and custom) won't respond without error to the console.
When being deployed, the 404 renders, but with the __NEXT_DATA__ json rendered as string to the dom.

When deployed, it's looking like:
![image](https://user-images.githubusercontent.com/9445140/146086208-02dc1cc5-d1bf-4489-8158-93e4d75108c2.png)

on reproducible url: https://next-rsc-demo-e3pu88wfg-klotzjesse.vercel.app/thisdoesnotexist

and console:
![image](https://user-images.githubusercontent.com/9445140/146086311-0bb96b4b-d59e-4acd-925e-7e881e75d457.png)

### Expected Behavior

Render in development.

And after deployment, without having __NEXT_DATA__ as string in the dom.

### To Reproduce

Update to canary 12.0.8-canary.5 up to .7
Activate in next.config.js
```
module.exports = {
  experimental: {
    reactRoot: true,
    concurrentFeatures: true,
    serverComponents: true,
  },
}
```

Deploy to Vercel / run `yarn dev`
Run into failure

Minimal Reproduction:
Failure (server): https://next-rsc-demo-e3pu88wfg-klotzjesse.vercel.app/rsc
Running (client): https://next-rsc-demo-e3pu88wfg-klotzjesse.vercel.app

Repo: https://github.com/KlotzJesse/next-rsc-demo/tree/rsc-fails-render-with-client-error

**Comment by KlotzJesse** [comment]
This render issues could somehow be related to #32514, i think? vice-versa

**Comment by KlotzJesse** [comment]
Still occuring on v12.0.8-canary.10:

Vercel reproduction deployment with v12.0.8-canary.10:
https://next-rsc-demo-klotzjesse.vercel.app/asd

I investigated:
Default error page won't render in dev, but on production.

Custom 404 will render in dev env, showing a different String.
![image](https://user-images.githubusercontent.com/9445140/146696497-fb6ca084-4d72-4660-acd5-691e48b2cd1a.png)



So I also found out some new errors in the local, dev env browser console:


![image](https://user-images.githubusercontent.com/9445140/146696613-7577c86e-de15-41be-8094-55b3b4674910.png)
![image](https://user-images.githubusercontent.com/9445140/146696630-30d5d681-3d58-47bb-84f0-942d464e4b4a.png)


After deployment it will show the same string as in the reproduction.
